### PR TITLE
fix(continuous-learning-v2): observer survives SIGUSR1 nudges without spurious failure

### LINKED ISSUE #1598: Observer logs spurious "Claude analysis failed (exit 158)" when SIGUSR1 fires during analysis
## Summary

`skills/continuous-learning-v2/agents/observer-loop.sh` reports `Claude analysis failed (exit 158)` and prematurely archives the observations file even when the Haiku analysis subprocess is still running and would have completed normally.

## Repro

1. Start an observer (any project with continuous-learning-v2 active).
2. Generate enough observations to trigger an analysis cycle.
3. While analysis is in flight, write another observation (which causes `hooks/observe.sh` to send SIGUSR1 to the observer).
4. Check the observer log.

Observed: `[Date] Claude analysis failed (exit 158)` followed by archival of the observations file.
Expected: Analysis completes normally and emits its actual exit code.

## Root cause

`observe.sh` sends SIGUSR1 to the observer to nudge it. Bash delivers the trapped signal to the in-flight `wait "$claude_pid"` call, which returns `128 + signum = 158` while the Haiku child is still running. The script logs failure and orphans the still-running subprocess.

This is documented bash behavior; there is no flag to disable signal-driven `wait` interruption.

## Fix

Re-enter `wait` until the child genuinely exits, gated by a `kill -0` liveness check. PR: #1417

### PR DESCRIPTION
**Pull request by AdamiecRadek**
Companion fix to #1431 (relocates data out of the path-guarded location).

## Summary

`observer-loop.sh` no longer treats a SIGUSR1-interrupted `wait` return (exit code > 128) as a Claude failure while the child process is still alive. The script re-enters `wait` until the child actually exits, then captures its real exit code.

Fixes #1598

## Type of Change
- [x] `fix:` Bug fix

## Testing Done
- [x] Manual testing completed
- [x] Automated tests pass locally (`node tests/run-all.js`)
- [x] Edge cases considered and tested

`tests/hooks/hooks.test.js` asserts the observer-loop source uses the resumable `while :; do ... wait` pattern and does not fall back to a bare `wait; exit_code=$?` capture. Two regression tests cover both the wait-loop pattern and the ANALYZING re-entrancy gate.

## Why

`hooks/observe.sh` sends SIGUSR1 to the observer whenever a new observation is logged. Bash delivers the trapped signal to the in-flight `wait` call, which exits with `128 + signum` (e.g. 158) even though the Haiku child is still running. The observer falsely logged `Claude analysis failed (exit 158)` and archived the observations file, while the orphan Haiku kept writing unscored.

## How it is fixed

```bash
exit_code=0
while :; do
  wait "$claude_pid"
  exit_code=$?
  if [ "$exit_code" -gt 128 ] && kill -0 "$claude_pid" 2>/dev/null; then
    continue
  fi
  break
done
```

`kill -0` is a non-destructive liveness check. The loop exits only when `wait` returns a genuine child exit code or the process is gone. The outer `while :;` (rather than `while kill -0`) ensures `wait` always runs at least once, so a fast non-zero exit cannot be masked by a pre-loop liveness check.

A second fix wraps the interval-driven `analyze_observations` call with `ANALYZING=1`/`ANALYZING=0`, so a SIGUSR1 trap firing mid-analysis hits the existing re-entrancy guard in `on_usr1()` instead of nesting and clobbering `claude_pid`/`watchdog_pid`.

## Security & Quality Checklist
- [x] No secrets or API keys committed
- [x] JSON files validate cleanly
- [x] Shell scripts pass shellcheck (modified file)
- [x] No sensitive data exposed in logs or output
- [x] Follows conventional commits format

## Documentation
- [x] Updated relevant documentation
- [x] Added comments for complex logic

## Scope

Originally this PR also added `--permission-mode acceptEdits` to the Haiku invocation, with the intent of fixing instinct writes. That change was reverted once empirical testing showed `acceptEdits` does not bypass Claude Code's hardcoded sensitive-path guard on `~/.claude/**` in `--print` mode. The real fix for instinct writes is relocating the homunculus data directory — see #1431.

## Summary by CodeRabbit

* **Bug Fixes**
  * Improved subprocess wait handling to avoid signal-induced false failures and preserve child exit results.
  * Signal handling now defers analysis triggers arriving during an in-flight analysis and runs them immediately afterward, preventing overlapping analyses and preserving analysis state.

* **Tests**
  * Added and updated tests to verify wait resilience, signal re-entrancy guards, and deferred-trigger behavior.

### CODE REVIEW

File: tests/hooks/hooks.test.js [hunk 1]
```diff
@@ -2441,6 +2441,27 @@ async function runTests() {
     passed++;
   else failed++;
 
+  if (
+    test('observer-loop runs the haiku analysis with --permission-mode acceptEdits', () => {
+      const observerLoopSource = fs.readFileSync(path.join(__dirname, '..', '..', 'skills', 'continuous-learning-v2', 'agents', 'observer-loop.sh'), 'utf8');
+
+      assert.ok(observerLoopSource.includes('--permission-mode acceptEdits'), 'observer-loop should spawn claude with --permission-mode acceptEdits so user settings.json plan-mode/Write allow-lists do not block instinct writes');
```
[the remark this answers is not in the document]
Comment (AdamiecRadek) [reply tests/hooks/hooks.test.js]: Addressed in dd33a5f. The assertion now anchors to the claude invocation (`/claude --model haiku[\s\S]*?--permission-mode acceptEdits[\s\S]*?-p "$prompt_content"/`), so a stray flag in a comment no longer satisfies it. Counter-tested: stripping the real flag while keeping the comment fails the test.

**Comment by AdamiecRadek** [comment]
Scope update: during post-merge validation I discovered that `--permission-mode acceptEdits` does **not** unblock instinct writes. Claude Code has a hardcoded sensitive-path guard on `~/.claude/**` that ignores every `--permission-mode` variant (including `bypassPermissions` and `--dangerously-skip-permissions`) in `--print` mode. Verified with isolated Haiku probes across all four modes.

Real root cause: `INSTINCTS_DIR` defaults to `~/.claude/homunculus/projects/<id>/instincts/personal/`, which is inside the guarded path. No CLI flag can fix that from the observer's subprocess.

**This PR is now amended** to carry only the SIGUSR1/`wait` resumable-loop fix. Removed `--permission-mode acceptEdits` and its test assertion so future readers aren't misled.

**Real fix** in #1431: relocate the observer homunculus dir to `${XDG_DATA_HOME:-$HOME/.local/share}/ecc-homunculus/` with `CLV2_HOMUNCULUS_DIR` override and a manual migration script.

**Comment by AdamiecRadek** [comment]
Addressed in 3624f27. Swapped the outer `while kill -0 "$claude_pid"` guard for `while :;` so `wait` always runs at least once. Fast non-zero exits are now captured instead of being masked by a pre-loop liveness check. The inner `[ $exit_code -gt 128 ] && kill -0` guard still handles SIGUSR1-induced resumption. Regression test in `tests/hooks/hooks.test.js` updated accordingly.

**Comment by AdamiecRadek** [comment]
@coderabbitai resume

**Comment by affaan-m** [comment]
Thanks for the SIGUSR1 observer-loop fix. I ported the signal deferral and wait-safety behavior onto current `main` in #1718, so I’m closing this as superseded.
